File: AsianCallFun.m
classdef AsianCallFun < fun
% §\mcommentfont Specify and generate payoff values of an Asian Call option$§
properties
   volatility = 0.5
   S0 = 30;
   K = 25;
   T = 1;
   A
   tVec
   dimFac = 1 
end
methods
   function obj = AsianCallFun(dimFac)
      if nargin
         dimVec = cumprod(dimFac);
         nf = numel(dimVec);
         obj(1,nf) = AsianCallFun;
         obj(1).dimFac = 0;      
         for ii = 1:nf
            d = dimVec(ii);
            if ii > 1
               obj(ii).dimFac = dimFac(ii-1);
            end
            obj(ii).dimension = d;
            tvec = (1:d)*(obj(ii).T/d);
            obj(ii).tVec = tvec;
            CovMat = min(tvec',tvec);
            [eigVec,eigVal] = eig(CovMat,'vector');
            obj(ii).A = sqrt(eigVal(end:-1:1)) .* eigVec(:,end:-1:1)';
         end
      end 
   end
   
   function y = f(obj, x, coordIndex)
      %since the nominalValue = 0, this is efficient
      BM = x * obj.A;
      SFine = obj.S0*exp((-obj.volatility^2/2)*obj.tVec + obj.volatility * BM);
      AvgFine = ((obj.S0/2) + sum(SFine(:,1:obj.dimension-1),2) + ...
         SFine(:,obj.dimension)/2)/obj.dimension;
      y = max(AvgFine - obj.K,0);
      if obj.dimFac > 0
         SCoarse = SFine(:,obj.dimFac:obj.dimFac:end);
         dCoarse = obj.dimension/obj.dimFac;
         AvgCoarse = ((obj.S0/2) + sum(SCoarse(:,1:dCoarse-1),2) + ...
            SCoarse(:,dCoarse)/2)/dCoarse;
         y = y - max(AvgCoarse - obj.K,0);
      end
   end
end
end

File: qmcpy/algorithms/integrand/AsianCallFun.py
from numpy import exp, maximum

from . import Integrand

class AsianCallFun(Integrand):
    ''' Specify and generate payoff values of an Asian Call option'''
    def __init__(self, BMmeasure=None, volatility=.5, S0=30, K=25, nominal_value=None):
        super().__init__(nominal_value=nominal_value)
        self.BMmeasure = BMmeasure
        self.volatility = volatility
        self.S0 = S0
        self.K = K
        self.dimFac = 0
        if self.BMmeasure:
            nBM = len(BMmeasure)
            self.fun_list = [AsianCallFun() for i in range(nBM)] 
            self[0].BMmeasure = self.BMmeasure[0]
            self[0].dimFac = 0
            self[0].dimension = self.BMmeasure[0].dimension
            for ii in range(1,nBM):
                self[ii].BMmeasure = self.BMmeasure[ii]
                self[ii].dimFac = self.BMmeasure[ii].dimension/self.BMmeasure[ii-1].dimension
                self[ii].dimension = self.BMmeasure[ii].dimension  

    def g(self,x,ignore):
        SFine = self.S0*exp((-self.volatility**2/2)*self.BMmeasure.timeVector+self.volatility*x)
        AvgFine = ((self.S0/2)+SFine[:,:self.dimension-1].sum(1)+SFine[:,self.dimension-1]/2)/self.dimension
        y = maximum(AvgFine-self.K,0)
        if self.dimFac > 0:
            Scourse = SFine[:,int(self.dimFac-1)::int(self.dimFac)]
            dCourse = self.dimension/self.dimFac
            AvgCourse = ((self.S0/2)+Scourse[:,:int(dCourse)-1].sum(1)+Scourse[:,int(dCourse)-1]/2)/dCourse
            y = y-maximum(AvgCourse-self.K,0)
        return y
